sky/clouds/vsphere.py

```diff
@@ -0,0 +1,325 @@
+from sky.provision.vsphere.vsphere_utils import initialize_vsphere_data
+
+if typing.TYPE_CHECKING:
+    # Renaming to avoid shadowing variables.
+    from sky import resources as resources_lib
+
+_CLOUD_VSPHERE = 'vsphere'
+_CREDENTIAL_FILES = [
+    # credential files for vSphere,
+    'credential.yaml',
+]
+
+
+@clouds.CLOUD_REGISTRY.register
+class Vsphere(clouds.Cloud):
+    """Vsphere cloud"""
+
+    _REPR = 'Vsphere'
+
+    _CLOUD_UNSUPPORTED_FEATURES = {
+        clouds.CloudImplementationFeatures.MULTI_NODE: 'Multi-node is not '
+                                                       'supported by the '
+                                                       'vSphere implementation '
+                                                       'yet.',
+        clouds.CloudImplementationFeatures.CLONE_DISK_FROM_CLUSTER:
+            (f'Migrating disk is currently not supported on {_REPR}.'),
+        clouds.CloudImplementationFeatures.DOCKER_IMAGE:
+            (f'Docker image is currently not supported on {_REPR}. '
+             'You can try running docker command inside the '
+             '`run` section in task.yaml.'),
+        clouds.CloudImplementationFeatures.SPOT_INSTANCE:
+            (f'Spot instances are not supported in {_REPR}.'),
+        clouds.CloudImplementationFeatures.CUSTOM_DISK_TIER:
+            (f'Custom disk tiers are not supported in {_REPR}.'),
+        clouds.CloudImplementationFeatures.OPEN_PORTS:
+            (f'Opening ports is currently not supported on {_REPR}.'),
+    }
+
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 80  # The name can't exceeds 80 characters
+    # for vsphere.
+
+    _regions: List[clouds.Region] = []
+
+    PROVISIONER_VERSION = clouds.ProvisionerVersion.SKYPILOT
+    STATUS_VERSION = clouds.StatusVersion.SKYPILOT
+
+    @classmethod
+    def _unsupported_features_for_resources(
+        cls, resources: 'resources_lib.Resources'
+    ) -> Dict[clouds.CloudImplementationFeatures, str]:
+        features = cls._CLOUD_UNSUPPORTED_FEATURES
+        if resources.use_spot:
+            features[clouds.CloudImplementationFeatures.STOP] = (
+                f'Stopping spot instances is currently '
+                f'not supported on {cls._REPR}.')
+        return features
+
+    @classmethod
+    def max_cluster_name_length(cls) -> Optional[int]:
+        return cls._MAX_CLUSTER_NAME_LEN_LIMIT
+
+    @classmethod
+    def regions(cls) -> List[clouds.Region]:
+        return service_catalog.regions(clouds=_CLOUD_VSPHERE)
+
+    @classmethod
+    def regions_with_offering(
+        cls,
+        instance_type: Optional[str],
+        accelerators: Optional[Dict[str, int]],
+        use_spot: bool,
+        region: Optional[str],
+        zone: Optional[str],
+    ) -> List[clouds.Region]:
+        del accelerators, zone  # unused
+        if use_spot:
+            return []
+        if instance_type is None:
+            # Fall back to default regions
+            regions = cls.regions()
+        else:
```

Comment:
The `instance_type` should be guaranteed to be specified by the optimizer, we don't have to do the if condition here.

Reply:
Removed